File: 2463-minimum-total-distance-traveled/solution.py
import math
from typing import List

class Solution:
    def minimumTotalDistance(self, robot: List[int], factory: List[List[int]]) -> int:
        # Sort robots and factories by position
        # Relative order is preserved in the optimal matching on a 1D line
        robot.sort()
        factory.sort()
        
        n, m = len(robot), len(factory)
        
        # memo[robot_idx][factory_idx] stores the min distance
        # to fix robots from robot_idx onwards using factories from factory_idx onwards
        memo = {}

        def solve(r_idx, f_idx):
            # Base Case: All robots are repaired
            if r_idx == n:
                return 0
            # Base Case: No more factories left, but robots remain
            if f_idx == m:
                return float('inf')
            
            state = (r_idx, f_idx)
            if state in memo:
                return memo[state]
            
            # Option 1: Skip the current factory
            res = solve(r_idx, f_idx + 1)
            
            # Option 2: Assign k robots to the current factory
            # k can range from 1 to factory[f_idx][1] (limit)
            curr_dist = 0
            limit = factory[f_idx][1]
            f_pos = factory[f_idx][0]
            
            for k in range(1, limit + 1):
                if r_idx + k > n:
                    break
                
                # Add distance for the k-th robot assigned to this factory
                curr_dist += abs(robot[r_idx + k - 1] - f_pos)
                
                # Recursive call for the next factory and remaining robots
                sub_res = solve(r_idx + k, f_idx + 1)
                if sub_res != float('inf'):
                    res = min(res, curr_dist + sub_res)
            
            memo[state] = res
            return res

        return solve(0, 0)

File: 2463-minimum-total-distance-traveled/solution.cpp
#include <vector>
#include <algorithm>
#include <cmath>

using namespace std;

class Solution {
    long long memo[101][101];
    const long long INF = 1e15; // Sufficiently large value for infinity

    long long solve(int robotIdx, int factoryIdx, const vector<int>& robot, const vector<vector<int>>& factory) {
        // Base case: All robots have been repaired
        if (robotIdx == robot.size()) return 0;
        // Base case: No more factories left, but robots remain
        if (factoryIdx == factory.size()) return INF;

        // Check memoization table
        if (memo[robotIdx][factoryIdx] != -1) return memo[robotIdx][factoryIdx];

        // Option 1: Skip the current factory entirely
        long long res = solve(robotIdx, factoryIdx + 1, robot, factory);

        // Option 2: Assign 'k' robots to the current factory
        long long currentDist = 0;
        int limit = factory[factoryIdx][1];
        int factoryPos = factory[factoryIdx][0];

        // Try assigning k robots, where 1 <= k <= factory limit and doesn't exceed remaining robots
        for (int k = 1; k <= limit && robotIdx + k <= robot.size(); ++k) {
            currentDist += abs(robot[robotIdx + k - 1] - factoryPos);
            long long subProblem = solve(robotIdx + k, factoryIdx + 1, robot, factory);
            
            if (subProblem != INF) {
                res = min(res, currentDist + subProblem);
            }
        }

        return memo[robotIdx][factoryIdx] = res;
    }

public:
    long long minimumTotalDistance(vector<int>& robot, vector<vector<int>>& factory) {
        // Sort both to ensure relative order is preserved
        sort(robot.begin(), robot.end());
        sort(factory.begin(), factory.end());

        // Initialize memo table with -1
        for (int i = 0; i <= robot.size(); ++i) {
            for (int j = 0; j <= factory.size(); ++j) {
                memo[i][j] = -1;
            }
        }

        return solve(0, 0, robot, factory);
    }
};